drivers/i2c/i2c_max32_rtio.c

```diff
@@ -0,0 +1,454 @@
+/*
+ * Copyright (c) 2024 Analog Devices, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT adi_max32_i2c
+
+#include <errno.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/i2c/rtio.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/clock_control/adi_max32_clock_control.h>
+#include <zephyr/irq.h>
+
+#include <wrap_max32_i2c.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(max32_i2c);
+
+#define ADI_MAX32_I2C_INT_FL0_MASK 0x00FFFFFF
+#define ADI_MAX32_I2C_INT_FL1_MASK 0x7
+
+#define ADI_MAX32_I2C_STATUS_MASTER_BUSY BIT(5)
+
+#define I2C_RECOVER_MAX_RETRIES 3
+#define I2C_STANDAR_BITRATE_CLKHI 0x12b
+
+static int complete_flag;
+
+/* Driver config */
+struct max32_i2c_config {
+	mxc_i2c_regs_t *regs;
+	const struct pinctrl_dev_config *pctrl;
+	const struct device *clock;
+	struct max32_perclk perclk;
+	uint32_t bitrate;
+#if defined(CONFIG_I2C_MAX32_INTERRUPT)
+	uint8_t irqn;
+	void (*irq_config_func)(const struct device *dev);
+#endif
+};
+
+struct max32_i2c_data {
+	mxc_i2c_req_t req;
+	const struct device *dev;
+	struct k_sem lock;
+	uint8_t target_mode;
+	uint8_t flags;
+	struct i2c_rtio *ctx;
+	uint32_t readb;
+	uint32_t written;
+	uint8_t second_msg_flag;
+#if defined(CONFIG_I2C_MAX32_INTERRUPT)
+	struct k_sem xfer;
+	int err;
+#endif
+};
+
+static int max32_configure(const struct device *dev,
+				uint32_t dev_cfg)
+{
+	struct i2c_rtio *const ctx = ((struct max32_i2c_data *)
+		dev->data)->ctx;
+
+	return i2c_rtio_configure(ctx, dev_cfg);
+}
+
+static int max32_do_configure(const struct device *dev, uint32_t dev_cfg)
+{
+	int ret = 0;
+	const struct max32_i2c_config *const cfg = dev->config;
+	mxc_i2c_regs_t *i2c = cfg->regs;
+
+	switch (I2C_SPEED_GET(dev_cfg)) {
+	case I2C_SPEED_STANDARD: /** I2C Standard Speed: 100 kHz */
+		ret = MXC_I2C_SetFrequency(i2c, MXC_I2C_STD_MODE);
+		break;
+
+	case I2C_SPEED_FAST: /** I2C Fast Speed: 400 kHz */
+		ret = MXC_I2C_SetFrequency(i2c, MXC_I2C_FAST_SPEED);
+		break;
+
+#if defined(MXC_I2C_FASTPLUS_SPEED)
+	case I2C_SPEED_FAST_PLUS: /** I2C Fast Plus Speed: 1 MHz */
+		ret = MXC_I2C_SetFrequency(i2c, MXC_I2C_FASTPLUS_SPEED);
+		break;
+#endif
+
+#if defined(MXC_I2C_HIGH_SPEED)
+	case I2C_SPEED_HIGH: /** I2C High Speed: 3.4 MHz */
+		ret = MXC_I2C_SetFrequency(i2c, MXC_I2C_HIGH_SPEED);
+		break;
+#endif
+
+	default:
+		/* Speed not supported */
+		return -ENOTSUP;
+	}
+
+	return ret;
+}
+
+static void max32_complete(const struct device *dev, int status);
+
+static int max32_msg_start(const struct device *dev, uint8_t flags,
+				 uint8_t *buf, size_t buf_len, uint16_t i2c_addr)
+{
+	int ret = 0;
+	const struct max32_i2c_config *const cfg = dev->config;
+	struct max32_i2c_data *data = dev->data;
+	mxc_i2c_regs_t *i2c = cfg->regs;
+	mxc_i2c_req_t *req = &data->req;
+	uint8_t target_rw;
+
+	req->i2c = i2c;
+	req->addr = i2c_addr;
+
+	k_sem_take(&data->lock, K_FOREVER);
```

Comment:
This semaphore should be unnecessary, i2c_rtio ctx, no locking is needed with an RTIO driver aside from the locking i2c_rtio ctx provides